[CMSP-695] WordPress 6.5 font library documentation and behavior (#8867)
pantheon-systems/documentation: source/content/guides/wordpress-configurations/07-wordpress-font-library.md

```php
add_filter( 'font_dir', function( $defaults ) {
	$font_dir = '/path/to/your/custom/dir';
	$font_url = site_url( $font_dir );
	
	$defaults['path'] = $font_dir;
	$defaults['url'] = $font_url;
	$defaults['basedir'] = $font_dir;
	$defaults['baseurl'] = $font_url;
	
	return $defaults;
} );
```

<Alert title="Note" type="info">

You cannot use the function `wp_get_upload_dir()` inside the `font_dir` filter because it will cause an infinite loop. Instead, if you want to use the `wp_get_upload_dir()` function, you can use a global variable as [we do in our mu-plugin](https://github.com/pantheon-systems/pantheon-mu-plugin/blob/main/inc/fonts.php) or [remove the filter and then re-add it](https://github.com/WordPress/wordcamp.org/pull/1245/files#diff-e441f1053cefcd468bd20fed91d1aac5e902871d7c564be909fc35590f9c3082R635-R637).

For more information, refer to [this Gutenberg issue](https://github.com/WordPress/gutenberg/issues/58696).

</Alert>

## Considerations

The distinction between the two approaches for handling fonts depends what works best for your workflow and whether you consider your fonts to be handled more like media files or more like themes and plugins.

> **If you consider fonts to be more like media files**, use the default Pantheon behavior added in our `mu-plugin`. This makes it easier to _install fonts onto production sites_ but perhaps more difficult to _synchronize_ those fonts on the lower environments (test and dev).

Reviewer: Using the default Pantheon filter, do fonts move between environments with a "clone files" workflow as expected?

Author: There are a few things I need to test. That's one. I also have a couple open questions that I posted in `#core-editor` Slack: https://wordpress.slack.com/archives/C02QB2JS7/p1707937998405009

Reviewer: I smashed that subscribe button for more jazzsequence gutenberg goodness.

Author: Here's what I'm seeing:

1. Install font on live site
2. Pull Live db down to Dev and sync files
3. The font appears in the font manager but is not _active_
4. Clicking into the font, selecting it and clicking update _activates_ it

So, to answer your question, it does appear that they are pulled down when you clone the DB and files (I suspect you would _need_ to clone the DB and files in order to get them), but there is an additional step to actually _use_ them on the lower environment you pulled down to. I will update the doc.

Reviewer: it's curious that activation status doesn't carry through db clone the way plugin activation status would.

Author: I described this somewhere, so I might be repeating myself, but I had a similar experience when I cloned a DB from live to Dev with theme styles. It kept everything else, but the styles that were active on Live were not active on Dev until I set them. I'm not sure what this tells me exactly, because surely they are stored in the database somewhere, but whatever the full site editor is doing with styles, it's doing the same thing with fonts. `¯\_(ツ)_/¯`

```php
add_filter( 'font_dir', function( $defaults ) {
	$font_dir = '/path/to/your/custom/dir';
	$font_url = site_url( $font_dir );
	
	$defaults['path'] = $font_dir;
	$defaults['url'] = $font_url;
	$defaults['basedir'] = $font_dir;
	$defaults['baseurl'] = $font_url;
	
	return $defaults;
} );
```

The default priority for WordPress filters is `10`. We have set _our_ `font_dir` filter to priority `9` so it allows you to override our modification.

<Alert title="Note" type="info">

You cannot use the function `wp_get_upload_dir()` inside the `font_dir` filter because it will cause an infinite loop. Instead, if you want to use the `wp_get_upload_dir()` function, you can use a global variable as [we do in our mu-plugin](https://github.com/pantheon-systems/pantheon-mu-plugin/blob/main/inc/fonts.php) or [remove the filter and then re-add it](https://github.com/WordPress/wordcamp.org/pull/1245/files#diff-e441f1053cefcd468bd20fed91d1aac5e902871d7c564be909fc35590f9c3082R635-R637).

For more information, refer to [this Gutenberg issue](https://github.com/WordPress/gutenberg/issues/58696).

</Alert>

## Considerations

> WordPress handles fonts more like **media files** than **plugins or themes**. This means that when fonts are added to one Pantheon site environment, they will not necessarily exist in your other WordPress environments (e.g. installing a font on Dev does not mean it will push to Test when you deploy). In WordPress, fonts have two parts, there is an underlying font post type (similar to the `attachment` post type for media files) and the physical font files themselves. Without the font post type existing in the database, WordPress has no way of knowing that a font is installed (in the same way that WordPress has no way that a particular image exists in the `/uploads` directory if it was not uploaded via the media library).

Reviewer: ```suggestion
WordPress handles fonts more like **media files** than **plugins or themes**. This means that when fonts are added to one Pantheon site environment, they will not necessarily exist in your other environments (e.g. installing a font on Dev does not mean it will push to Test when you deploy). In WordPress, fonts have two parts, there is a font post type (similar to the `attachment` post type for media files) and the font files themselves. Without the font post type existing in the database, WordPress has no way of knowing that a font is installed (in the same way that WordPress has no way that a particular image exists in the `/uploads` directory if it was not uploaded via the media library).
```

> If you intend to override the Pantheon behavior, care should be taken that you are not doing so with the assumption that committing font files to your repository will make them available automatically on your other environments.

Reviewer: ```suggestion
If you intend to override the Pantheon behavior, know that committing font files to your repository will not make them available automatically on your other environments.
```

### I've cloned my database from Live and my font is installed but not "active"

![Font is installed but not active](../../../images/wordpress-configurations/07-font-installed-not-active.png)

If you've cloned your database and files from your live environment to Dev or Test and the font appears in your Font Library but is not "active" (it displays a message like "0/1 variants active"), you can click into the font, select the variant you want, and click update. 

![Update Font](../../../images/wordpress-configurations/07-font-update.png)

Whether or not a font is "active" is similar to whether or not a plugin or theme is active. If a font is not active, it means the font is installed and recognized by WordPress but it cannot be used on the site until you activate it.

### I've uploaded fonts to my `/fonts` directory, but they aren't showing up in the Font Library

Fonts need to be installed via the Font Library in the WordPress admin. This is because fonts contain data that is stored in a post type. Simply having a font in your `/fonts` directory does not mean it will be recognized by WordPress.

## More Resources

> * [Font Library Gutenberg Tracking Ticket](https://github.com/WordPress/gutenberg/issues/55277)

Reviewer: Are there any WordPress core docs that could be linked here?

Author: No, not yet. Otherwise they would be linked.

1. The site environment is Dev and is in Git mode. If the Dev environment is set to Git mode, the filesystem is not writeable and you will not be able to install fonts.
2. The site environment is Test or Live and you've set the `'pantheon_modify_fonts_dir'` filter value to `false`. In this case, you will not be able to install fonts on any environment other than Dev and you will need to commit them to your codebase.

### I've cloned my database from Live and my font is installed but not "active"

![Font is installed but not active](../../../images/wordpress-configurations/07-font-installed-not-active.png)

If you've cloned your database and files from your live environment to Dev or Test and the font appears in your Font Library but is not "active" (it displays a message like "0/1 variants active"), you can click into the font, select the variant you want, and click update. 

![Update Font](../../../images/wordpress-configurations/07-font-update.png)

### I've uploaded fonts to my `/fonts` directory, but they aren't showing up in the Font Library

> Fonts need to be installed via the Font Library in the WordPress admin. This is because, in the backend, they are stored as a post type. Therefore, simply having a font in your `/fonts` directory does not necessarily mean it will be recognized by WordPress.

Reviewer: Isn't this the state if we're tracking fonts in version control and pushing up "as code" then? What is the next step then?

Author: Good point. I guess you'd need to install it via the admin (or in code) again? I'm still waiting for clarification in WP core Slack because the other option would be cloning the DB from a lower environment. I tried exporting a WXR and importing it into a local copy of the same site and I got a Gutenberg error when I tried to open the font library, so that's a no-go. When you upload a font into the Font Library where the files already exist, you get an error (that a font with that name already exists) but it _does_ show up in the font library. 

There's an open ticket dealing with fonts still being listed after the _files_ have been deleted: https://github.com/WordPress/gutenberg/issues/58375

There's also an open ticket for exporting fonts in a WP Export: https://github.com/WordPress/gutenberg/issues/54226 -- this one might be the missing next step, particularly if you could export _just_ fonts.

There isn't (yet) a ticket dealing with font _files_ existing but I thought I remembered seeing a discussion about it. I will open an issue since my Slack thread hasn't gotten any attention.

Author: https://github.com/WordPress/gutenberg/issues/59102

**If you consider fonts to be more like themes and plugins**, use the WordPress default behavior and the `pantheon_modify_fonts_dir` filter. This ensures that the fonts are installed across environments, but does not necessarily mean that fonts that are _physically installed_ on your lower environments (e.g. they exist in the `wp-content/fonts` directory) are recognized by WordPress. (They may still need to be added in the WordPress admin.)

## Troubleshooting

### "No font faces were installed" error message

![No font faces were installed]((../../../images/wordpress-configurations/07-no-font-faces-installed.png))

> There are two reasons you might receive an error message like the above that says that no fonts were installed and asks if the parent directory is writable by the server, both assume you have _disabled_ our modification to the directory where fonts are installed and are using the WordPress core handling (uploading to `wp-content/fonts/`).

Reviewer: ```suggestion
If you have disabled our modification and are using WordPress's default handling (uploading to `wp-content/fonts/`), there are two reasons you might receive an error message reporting that no fonts were installed and asks if the parent directory is writable by the server.
```
Be more clear with "this only happens if you've changed the Pantheon setting"
